### ISSUE TITLE
Uniformity

### ISSUE THREAD

**Issue by vladimirias** [UniMath/UniMath#687]
We need uniformity in naming. It is crucial.

1. Now we have four different ways of forming a name from two (or more) words. Let's say monoidfun can be named as:

monoidfun
monoid_fun
MonoidFun
Monoid_Fun

we *must* for the integrity of the library choose one of these four and apply it to *all* names in the library.

Those who want to preserve there own preferred way of doing it if it differs from the UniMath one will have to develop his own own UniMath-based library in his own GitHub account.

I suggest that we avoid capitalization as it makes it difficult to remember a composite name as one word and choose one of the first two.

**Comment by DanGrayson** [comment]
This is a good place for me to advertise my past proposal: #176.

An advantage of capitalization is that it makes it possible to separate an identifier into words, especially where the words are unrecognizable to English speakers, such as in "coconustot".

I think we should also be uniform in *never* including arbitrary parameter names in identifier names, such as the "t" in coconustot, since there is no way to guess whether it is coconustot or coconustox or coconustoy, etc.

By the way, why is there "coconustot" but not "conustot"?

**Comment by siddharthist** [comment]
@DanGrayson I like your proposal, and think it should be implemented whenever possible. I also think that UniMath should have a Travis script that checks for conformity. Would you be willing to write up these guidelines in `STYLE.md` so future contributors can refer to it? Or is there some level of consensus that should be reached first?

**Comment by peterlefanulumsdaine** [comment]
I agree it would be good to have a guideline for this.  General conventions on identifier naming (as discussed in Dan’s #176) would be great, but are very hard to design well — but “how to concatenate words” is a specific comparatively simple one that’s easy to decide on.

Following @DanGrayson above, I would argue strongly against the form `monoidfun` — i.e. concatenation with *no* marker of word boundaries.  These are really just unnecessarily difficult to read.

Between the rest, my preference isn’t so strong.  But overall, I vote for the form `monoid_fun` — i.e. always use underscores for concatenation of words, without usually capitalising them.  This seems visually clear, plus it leaves capitalisation free to be either avoided entirely, or used for some other distinction.  (E.g. a convention like “plain words are never capitalised; acronyms and single-letter names may be capitalised iff they would be in traditional usage”.)

Whatever form is preferred, should this be (1) an absolute rule, no exceptions allowed; (2) a strong guideline (we ask contributors to justify exceptions, unless the justification is obvious); or (3) a recommendation (contributors may make other choices if they prefer to)?  I vote for (2) or (3) — not (1), since there will probably be occasional good reasons to make exceptions which we don’t foresee now.

**Comment by DanGrayson** [comment]
Yes, `monoidfun` is a good example, for it could be "monoid fun" or "mono id fun".

**Comment by DanGrayson** [comment]
It's important to reach a consensus before we instate the official rule.  Making all of our code instantly comply with whatever convention we agree upon is probably not worth it, so the convention might be thought of a mandatory for new code and recommended, in the long run, for old code.

**Comment by varkor** [comment]
As someone without a huge amount of experience with UniMath, I can say that the current naming convention (i.e. often very terse and nonconformant) is a large hurdle when trying to learn one's way around the code base. When encountering names, a lot of the time it is not clear what they are for, as the names are unreadable (e.g. `coconustot`, `total2asstor`, `weqii2withneg`,`ezmappr1` to pick out a few from `PartA.v`). With some practice, the contractions become more familiar, but it is still difficult. (Often, even looking at the definitions doesn't help, because they make use of yet more unfamiliar contractions.)

On top of this, the names are not consistent (e.g. whether to abbreviate `precategory` to `precat` or `algebra` to `alg` and so on).

UniMath would be significantly more pleasant to work with and contribute to if a consistent and readable naming scheme was decided upon (and implemented!).

---

> It's important to reach a consensus before we instate the official rule.

This issue has been open for over a year with no recent discussion and no change as far as I can see. https://github.com/UniMath/UniMath/issues/176 has been open even longer. I think it's easy to underestimate how important an issue this is, as most of those actively working on UniMath are familiar with the code and presumably do not have so much trouble navigating it as a newcomer.

---

> Making all of our code instantly comply with whatever convention we agree upon is probably not worth it

I think it would be worth the effort to try to rename existing definitions using a new naming convention to make the code base more accessible to newcomers. I think the time spent refactoring will be paid for by the contributors who are not put off by the difficulty of the current code.

**Comment by varkor** [comment]
For what it's worth, my preference of style is similar to @DanGrayson's in https://github.com/UniMath/UniMath/issues/176.

- Names should not be abbreviated unless **commonly/traditionally** done so in mathematical writing (i.e. someone using such a definition/theorem would be expected to understand the abbreviation immediately) --- for example `ob` for `object`, `iso` for isomorphism, `pr1` for the first projection and so on.
	- However, abbreviations should be consistently applied.
- Names should be fragments of English (e.g. `has_property`, `is_structure`, `pr1_preserves_limit`).
- Definitions (e.g. of structures) should be capitalised in camel case (e.g. `Precategory`, `FunctorAlgebra`); whereas theorems/lemmas/etc. should be lowercase in snake case (e.g. `pentagon_identity`, `identity_is_epi`).
- It is acceptable to *locally* abbreviate names as long as these do not escape and it is easy to find the full name within the file.
- Any exceptions to these conventions/rules *must* include a comment directly above the definition explaining the definition/theorem/etc. (Comments are sorely lacking from UniMath at the moment.)

Some points of discussion:
- For names, with apostrophes or diacritics, a point could be made either for, for example, `adameks_theorem`, something more accurate such as `Adámek's_theorem` or something in between. (My preference is towards the latter, as autocomplete should make it easy to type anyway, but I could see the argument for a simpler name.)
- At the moment, there are many trailing names (e.g. `is_precategory_X` rather than `X_is_precategory`). Personally, for readability, I would prefer the latter, but perhaps there's a reason the current convention was chosen?

**Comment by DanGrayson** [comment]
@varkor: What is your name?  It would be great if you would install it and a nice portrait of you in your github profile so we could get to know you better.

**Comment by DanGrayson** [comment]
Re: Peter's recommendation: "always use underscores for concatenation of words, without usually capitalising them":

This might actually be better than avoiding underscores, as I have proposed, because that forces capitalization, which is not always appropriate for identifiers and for words like "is".

I still sort of like the idea of making the first letter of structural types uppercase, e.g., Ring, Group, Abelian_group, Additive_category, Kernel, Cokernel, Category, etc., and letting the first letter of functions and predicates be lower case: is_a_proposition, is_a_set, is_a_kernel.  On the other hand, Vladimir spoke against it in opening this issue.  Then we would have such things as ring, group, abelian_group, additive_category, kernel, cokernel, category, etc.  One positive point about never capitalizing is that searching for words will be more fruitful, e.g., searching for "category" will not be problematic because there will be no uses of "Category".

**Comment by siddharthist** [comment]
Not to reopen our favorite old can of worms... but has anyone expressed opposition to @DanGrayson's proposal(s) for naming? It appears to me that Peter, Dan, @varkor and I have reached a consensus on at least the following: we should use underscores to demarcate word boundaries. 

Can we add that guideline alone to the styleguide, and improve upon its specificity later?

**Comment by benediktahrens** [comment]
On 21/04/2019 21:49, Langston Barrett wrote:
> Not to reopen our favorite old can of worms... but has anyone expressed 
> opposition to @DanGrayson <https://github.com/DanGrayson>'s proposal(s) 
> for naming? It appears to me that Peter, Dan, @varkor 
> <https://github.com/varkor> and I have reached a consensus on at least 
> the following: we should use underscores to demarcate word boundaries.
> 
> Can we add that guideline alone to the styleguide, and improve upon its 
> specificity later?

Yes, that's fine with me.

**Comment by DanGrayson** [comment]
I'll make a pull request to that effect.

**Comment by arnoudvanderleer** [comment]
It seems that currently there is an emerging "standard" to have all identifiers written in `snake_case`. See, for example, #1814 , #1831 (, #1768 ). Would it be an idea to make this the formal standard?
